pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,364 @@
+    if an input :math:`x_j, j = 1 \dots N`,
+    only enters into single-parameter gates of the form :math:`e^{-i x_j G}`
+    (where :math:`G` is a Hermitian generator),
+    the frequency spectrum :math:`\Omega_j` is fully determined by the eigenvalues
+    of the generators :math:`G`. In many situations, the spectra are limited
+    to a few frequencies only, which in turn limits the function class that the circuit
+    can express.
+
+    The ``advanced_spectrum`` function computes all frequencies that will
+    potentially appear in the sets :math:`\Omega_1` to :math:`\Omega_N`.
+
+    .. note::
+
+        In more detail, the ``advanced_spectrum`` function also allows for
+        preprocessing of the QNode arguments before they are fed into the gates,
+        as long as this processing is *linear*. In particular, constant
+        prefactors for the encoding arguments are allowed.
+
+    **Example**
+
+    Consider the following example, which uses non-trainable inputs ``x``, ``y`` and ``z``
+    as well as trainable parameters ``w`` as arguments to the QNode.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from advanced_spectrum import advanced_spectrum
+
+        n_qubits = 3
+        dev = qml.device("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev)
+        def circuit(x, y, z, w):
+            for i in range(n_qubits):
+                qml.RX(0.5*x[i], wires=i)
+                qml.Rot(w[0,i,0], w[0,i,1], w[0,i,2], wires=i)
+                qml.RY(2.3*y[i], wires=i)
+                qml.Rot(w[1,i,0], w[1,i,1], w[1,i,2], wires=i)
+                qml.RX(z, wires=i)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = np.array([1., 2., 3.])
+        y = np.array([0.1, 0.3, 0.5])
+        z = -1.8
+        w = np.random.random((2, n_qubits, 3))
+        res = advanced_spectrum(circuit, argnum=[0, 1, 2])(x, y, z, w)
+
+    This circuit looks as follows:
+
+    >>> print(qml.draw(circuit)(x, y, z, w))
+    0: ──RX(0.5)──Rot(0.598, 0.949, 0.346)───RY(0.23)──Rot(0.693, 0.0738, 0.246)──RX(-1.8)──┤ ⟨Z⟩
+    1: ──RX(1)────Rot(0.0711, 0.701, 0.445)──RY(0.69)──Rot(0.32, 0.0482, 0.437)───RX(-1.8)──┤
+    2: ──RX(1.5)──Rot(0.401, 0.0795, 0.731)──RY(1.15)──Rot(0.756, 0.38, 0.38)─────RX(-1.8)──┤
+
+    Applying the ``advanced_spectrum`` function to the circuit for
+    the non-trainable parameters, we obtain:
+
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "x": {(0,): [-0.5, 0.0, 0.5], (1,): [-0.5, 0.0, 0.5], (2,): [-0.5, 0.0, 0.5]}
+    "y": {(0,): [-2.3, 0.0, 2.3], (1,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+    "z": {(): [-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0]}
+
+    .. note::
+        While the Fourier spectrum usually does not depend
+        on trainable circuit parameters or the actual values of the inputs,
+        it may still change based on inputs to the QNode that alter the architecture
+        of the circuit.
+
```

Comment:
It might be useful to have an example using the other spectrum function on the same or a similar circuit to highlight the difference in functionality.

Reply:
Very good call!

Reply:
Added one :)